In [1]:
import torch
import csv
import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import interpolate

In [6]:
def cal_valid_time(t_data, interval=72):
    state = 'invalid'
    state_count = 0
    start_t = []
    end_t = []
    for i, d in enumerate(t_data):
        if d != -99999:
            if state == 'invalid':
                state = 'valid'
                start_t.append(i)              
            state_count = 0
        else:
            if state == 'invalid':
                continue
            else:
                if state_count >= interval:
                    state = 'invalid'
                    state_count = 0
                    end_t.append(i - interval - 1)
                else:
                    state_count += 1
    if len(start_t) > len(end_t):
        end_t.append(len(t_data) - 1)
    return start_t, end_t


def generate_val_data(t_data, interval=72):
    start_t = []
    end_t = []
    count = 0
    state = 'invalid'
    flag = True
    for i, d in enumerate(t_data):
        if d != -99999:
            if state == 'invalid':
                state = 'valid'
            elif count >= interval and flag == True:
                start_t.append(i - interval)
                flag = False
            count += 1
        else:
            if flag == False:
                flag = True
                end_t.append(i - 1)
            count = 0
    if len(start_t) > len(end_t) and flag == False:
        end_t.append(len(data) - 1)
    return start_t, end_t
        

In [18]:
dt = pd.read_csv("AQ_NO2-19800101-20220504.csv", header=4, dtype=str)
# print(dt)
dt = dt.values
i = 0
for item in dt:
#     print(item)
    i += 1
    if i > 5:
        break


In [26]:
for index in range(3, 18):
    station_ID = dt[0, index+1]
    print(str(index+1)+'.', ' ID: ', station_ID, '  ', sep='')
    print('Location: ', dt[1, index+1], dt[2, index+1], '  ')
    station_data = dt[4:, index+1].astype(float)
    t = dt[4:, 0]
    s_t, e_t = cal_valid_time(station_data)
    # print(s_t, e_t)
    print('start_time: ', [t[i] for i in s_t], '  ')
    print('end_time: ', [t[i] for i in e_t], '  ')

    valid_data = list(station_data[s_t[0]:])
    print('invalid_times: ', valid_data.count(-99999), '  ratio:', list(valid_data).count(-99999)/len(valid_data), '  ')
    print("<img src='" + station_ID.split('_')[0] + ".png'> ")

4. ID: EN_A  
Location:  22.2845 114.2169   
start_time:  ['1999/1/1 00:00', '1999/10/4 12:00', '2002/3/6 17:00', '2002/3/15 16:00', '2008/6/10 12:00', '2009/7/27 18:00']   
end_time:  ['1999/5/25 21:00', '2002/2/4 15:00', '2002/3/10 20:00', '2008/6/7 05:00', '2009/7/23 17:00', '2022/5/4 23:00']   
invalid_times:  12162   ratio: 0.05944281524926686   
<img src='EN.png'> 
5. ID: KC_A  
Location:  22.3586 114.1271   
start_time:  ['1988/9/1 01:00', '1988/9/27 14:00', '1991/6/25 14:00', '1992/5/12 13:00', '1992/6/16 13:00', '1992/6/23 13:00', '1992/7/9 12:00', '1993/3/2 12:00', '1993/12/15 17:00', '1994/5/11 03:00', '1994/11/29 15:00', '1995/2/7 15:00', '1996/2/22 14:00', '1997/6/11 03:00', '1997/12/2 19:00', '2008/7/2 18:00', '2013/1/14 17:00']   
end_time:  ['1988/9/24 05:00', '1991/6/21 16:00', '1992/5/8 14:00', '1992/6/12 08:00', '1992/6/19 11:00', '1992/7/4 09:00', '1993/2/22 09:00', '1993/12/10 16:00', '1994/5/7 16:00', '1994/11/25 22:00', '1995/1/23 16:00', '1996/2/18 22:00', '1997

1. ID: CB_R  （road）  
Location:  22.2819 114.1822   
start_time:  ['1997/12/18 00:00', '1998/2/2 13:00', '2002/12/17 01:00', '2005/6/3 17:00', '2008/12/4 16:00']   
end_time:  ['1998/1/28 07:00', '2002/10/24 00:00', '2005/4/18 11:00', '2008/12/1 10:00', '2022/5/4 23:00']   
invalid_times:  8264   ratio: 0.038671758011380654   
<img src='CB.png'> 
2. ID: CL_R  （road）  
Location:  22.2833 114.1557   
start_time:  ['1998/7/12 00:00', '1998/11/13 12:00', '1998/12/31 16:00', '2000/8/28 17:00']   
end_time:  ['1998/11/5 11:00', '1998/12/4 23:00', '2000/8/25 13:00', '2022/5/4 23:00']   
invalid_times:  5321   ratio: 0.025489576147773436   
<img src='CL.png'> 
3. ID: CW_A  
Location:  22.2868 114.1429   
start_time:  ['1983/11/14 16:00', '1984/7/11 16:00', '1984/8/22 16:00', '1985/3/10 14:00', '1985/3/19 12:00', '1986/1/27 18:00', '1986/4/17 13:00', '1986/6/27 11:00', '1986/7/14 16:00', '1986/8/12 12:00', '1987/2/6 12:00', '1990/7/1 02:00', '1992/7/1 14:00', '1993/5/3 13:00', '1994/11/24 03:00', '1995/1/6 21:00', '1995/3/28 19:00', '1995/6/12 21:00', '1997/5/24 03:00', '1997/12/24 01:00', '2002/3/18 11:00']   
end_time:  ['1984/7/5 09:00', '1984/8/17 12:00', '1985/3/4 09:00', '1985/3/16 05:00', '1986/1/22 11:00', '1986/4/9 13:00', '1986/6/13 23:00', '1986/6/30 23:00', '1986/7/31 13:00', '1987/1/27 07:00', '1990/6/11 00:00', '1992/2/14 10:00', '1993/4/23 00:00', '1994/11/10 00:00', '1994/12/19 00:00', '1995/3/21 11:00', '1995/6/9 13:00', '1997/5/19 11:00', '1997/12/4 11:00', '2001/11/13 10:00', '2022/5/4 23:00']   
invalid_times:  23399   ratio: 0.06938546757128623   
<img src='CW.png'> 
4. ID: EN_A  
Location:  22.2845 114.2169   
start_time:  ['1999/1/1 00:00', '1999/10/4 12:00', '2002/3/6 17:00', '2002/3/15 16:00', '2008/6/10 12:00', '2009/7/27 18:00']   
end_time:  ['1999/5/25 21:00', '2002/2/4 15:00', '2002/3/10 20:00', '2008/6/7 05:00', '2009/7/23 17:00', '2022/5/4 23:00']   
invalid_times:  12162   ratio: 0.05944281524926686   
<img src='EN.png'> 
5. ID: KC_A  
Location:  22.3586 114.1271   
start_time:  ['1988/9/1 01:00', '1988/9/27 14:00', '1991/6/25 14:00', '1992/5/12 13:00', '1992/6/16 13:00', '1992/6/23 13:00', '1992/7/9 12:00', '1993/3/2 12:00', '1993/12/15 17:00', '1994/5/11 03:00', '1994/11/29 15:00', '1995/2/7 15:00', '1996/2/22 14:00', '1997/6/11 03:00', '1997/12/2 19:00', '2008/7/2 18:00', '2013/1/14 17:00']   
end_time:  ['1988/9/24 05:00', '1991/6/21 16:00', '1992/5/8 14:00', '1992/6/12 08:00', '1992/6/19 11:00', '1992/7/4 09:00', '1993/2/22 09:00', '1993/12/10 16:00', '1994/5/7 16:00', '1994/11/25 22:00', '1995/1/23 16:00', '1996/2/18 22:00', '1997/6/6 00:00', '1997/11/11 11:00', '2008/6/27 09:00', '2013/1/11 09:00', '2022/5/4 23:00']   
invalid_times:  13703   ratio: 0.04642330820699585   
<img src='KC.png'> 
6. ID: KT_A  
Location:  22.3147 114.2233   
start_time:  ['1984/7/13 16:00', '1984/8/20 13:00', '1986/1/24 19:00', '1986/4/11 17:00', '1986/6/17 15:00', '1986/7/7 14:00', '1987/7/21 11:00', '1989/7/24 15:00', '1993/1/30 12:00', '1993/3/19 02:00', '1993/8/4 16:00', '1994/11/17 14:00', '1994/12/30 03:00', '1997/7/25 03:00', '1999/1/12 18:00', '2000/2/3 13:00', '2000/4/12 16:00', '2002/11/1 02:00', '2017/2/11 12:00']   
end_time:  ['1984/8/14 08:00', '1986/1/19 10:00', '1986/4/7 12:00', '1986/6/14 05:00', '1986/6/30 21:00', '1987/7/17 23:00', '1989/5/9 20:00', '1993/1/26 00:00', '1993/3/10 00:00', '1993/8/1 00:00', '1994/11/14 09:00', '1994/12/27 00:00', '1997/7/7 12:00', '1998/11/18 11:00', '2000/1/25 11:00', '2000/2/8 16:00', '2002/4/25 14:00', '2017/2/6 22:00', '2022/5/4 23:00']   
invalid_times:  23958   ratio: 0.07228806604229024   
<img src='KT.png'> 
7. ID: MKaR  (road)  
Location:  22.324 114.166   
start_time:  ['1991/4/1 02:00', '1992/5/1 02:00', '1994/6/1 17:00', '1995/4/7 19:00', '1995/4/19 03:00', '1995/6/21 20:00', '1996/1/23 20:00', '1996/7/20 14:00', '1997/2/12 13:00', '1997/7/12 18:00', '1997/11/18 14:00', '1998/5/20 02:00', '2006/11/13 15:00', '2011/8/12 21:00']   
end_time:  ['1992/4/26 00:00', '1994/5/27 00:00', '1995/4/1 00:00', '1995/4/15 07:00', '1995/6/17 00:00', '1996/1/20 00:00', '1996/7/17 12:00', '1997/2/8 21:00', '1997/6/16 11:00', '1997/11/15 02:00', '1998/5/15 11:00', '2006/11/6 11:00', '2011/8/2 16:00', '2022/5/4 23:00']   
invalid_times:  13643   ratio: 0.05005393189172531   
<img src='MKaR.png'> 
8. ID: NH_A  (skip)  
Location:  22.4968 114.1284   
start_time:  ['2020/7/10 10:00']   
end_time:  ['2022/5/4 23:00']   
invalid_times:  249   ratio: 0.015634811000879066   
9. ID: ST_A  
Location:  22.378 114.182   
start_time:  ['1994/4/28 18:00', '1996/7/9 19:00', '1997/9/13 03:00', '2007/1/25 02:00', '2011/5/4 18:00', '2012/5/7 14:00', '2016/10/25 19:00']   
end_time:  ['1996/7/4 04:00', '1997/8/27 10:00', '2006/11/17 12:00', '2011/4/30 09:00', '2012/4/24 11:00', '2016/10/18 17:00', '2022/5/4 23:00']   
invalid_times:  9970   ratio: 0.040594793117207796   
<img src='ST.png'> 
10. ID: SP_A  
Location:  22.3315 114.1567   
start_time:  ['1989/1/1 00:00', '1989/9/30 15:00', '1989/10/29 07:00', '1989/11/14 13:00', '1990/11/30 17:00', '1992/6/19 14:00', '1992/9/8 15:00', '1992/10/3 13:00', '1993/9/1 02:00', '1994/2/2 15:00', '1994/4/21 14:00', '1995/8/29 19:00', '1995/11/2 18:00', '1995/12/1 02:00', '1995/12/28 15:00', '1996/5/1 18:00', '1996/9/27 14:00', '1998/1/23 02:00', '1999/5/19 19:00', '2004/3/9 02:00']   
end_time:  ['1989/9/25 21:00', '1989/10/21 16:00', '1989/11/8 12:00', '1990/11/17 19:00', '1992/6/16 00:00', '1992/6/22 10:00', '1992/9/25 13:00', '1993/8/1 00:00', '1993/12/3 10:00', '1994/4/13 14:00', '1995/8/21 12:00', '1995/10/30 13:00', '1995/11/27 15:00', '1995/12/23 23:00', '1996/4/15 13:00', '1996/9/13 00:00', '1998/1/2 10:00', '1999/5/16 06:00', '2003/12/5 08:00', '2022/5/4 23:00']   
invalid_times:  18253   ratio: 0.06245722810763461   
<img src='SP.png'> 
11. ID: SN_A  (skip)  
Location:  22.2475 114.16   
start_time:  ['2020/7/10 10:00']   
end_time:  ['2022/5/4 23:00']   
invalid_times:  237   ratio: 0.014881326133366821   
12. ID: TP_A  
Location:  22.4524 114.162   
start_time:  ['1990/2/12 14:00', '1990/9/18 15:00', '1991/5/20 13:00', '1992/11/30 15:00', '1995/1/18 20:00', '1996/2/16 19:00', '1997/10/9 03:00', '1999/6/7 19:00', '1999/12/9 11:00', '2000/3/24 16:00', '2000/10/19 16:00', '2003/2/12 20:00', '2004/9/3 18:00', '2005/1/4 10:00', '2006/2/14 01:00', '2009/1/22 14:00', '2010/7/14 18:00']   
end_time:  ['1990/9/14 11:00', '1991/5/8 11:00', '1992/11/25 13:00', '1995/1/12 11:00', '1996/2/13 00:00', '1997/9/18 12:00', '1999/6/4 13:00', '1999/11/29 22:00', '2000/3/9 15:00', '2000/10/15 02:00', '2003/2/7 01:00', '2004/4/16 15:00', '2004/11/15 09:00', '2005/12/21 12:00', '2008/12/30 23:00', '2010/6/27 09:00', '2022/5/4 23:00']   
invalid_times:  21130   ratio: 0.07480546331239867   
<img src='TP.png'> 
13. ID: MB_A  
Location:  22.4728 114.3583   
start_time:  ['1998/3/13 21:00', '1998/4/8 20:00', '1998/11/4 17:00', '1999/6/1 01:00', '2004/9/15 15:00', '2010/4/14 20:00', '2010/11/22 14:00', '2016/2/26 10:00', '2018/9/27 13:00']   
end_time:  ['1998/4/4 00:00', '1998/10/30 13:00', '1999/5/1 00:00', '2004/7/1 00:00', '2010/4/8 00:00', '2010/11/18 11:00', '2015/11/30 11:00', '2018/9/16 10:00', '2022/5/4 23:00']   
invalid_times:  14928   ratio: 0.0705365369622227   
<img src='MB.png'> 
14. ID: TK_A  (skip)  
Location:  22.3177 114.2594   
start_time:  ['2016/3/16 10:00']   
end_time:  ['2022/5/4 23:00']   
invalid_times:  1359   ratio: 0.02527243649347268   
15. ID: TW_A  
Location:  22.3733 114.1121   
start_time:  ['1988/8/1 01:00', '1993/6/4 18:00', '1993/11/18 17:00', '1993/12/3 16:00', '1994/1/3 13:00', '1995/7/5 19:00', '1996/1/14 03:00', '1996/7/15 16:00', '1997/5/7 21:00', '1997/8/27 03:00', '1999/4/10 17:00', '2003/8/21 02:00', '2008/11/11 20:00']   
end_time:  ['1993/1/9 10:00', '1993/11/15 11:00', '1993/11/26 00:00', '1993/12/31 03:00', '1995/7/2 00:00', '1996/1/8 15:00', '1996/7/12 13:00', '1997/4/21 11:00', '1997/7/29 11:00', '1999/4/7 04:00', '2003/2/8 04:00', '2008/11/8 15:00', '2022/5/4 23:00']   
invalid_times:  22658   ratio: 0.0765682500954653   
<img src='TW.png'> 
16. ID: TM_A  (skip)  
Location:  22.3908 113.9767   
start_time:  ['2013/12/29 17:00', '2014/12/31 14:00', '2018/12/17 18:00']   
end_time:  ['2014/11/3 09:00', '2018/12/13 14:00', '2022/5/4 23:00']   
invalid_times:  2488   ratio: 0.034008119301794724   
17. ID: TC_A  
Location:  22.2903 113.9411   
start_time:  ['1990/1/17 15:00', '1990/3/27 13:00', '1999/3/18 00:00', '2005/9/5 13:00']   
end_time:  ['1990/3/24 08:00', '1993/7/31 23:00', '2005/9/2 11:00', '2022/5/4 23:00']   
invalid_times:  59043   ratio: 0.20856691711793818   
<img src='TC.png'> 
18. ID: YL_A  
Location:  22.4467 114.0203   
start_time:  ['1993/12/10 14:00', '1995/8/1 03:00', '1997/5/13 03:00', '1997/7/4 17:00', '1997/8/27 19:00', '2002/1/16 20:00', '2002/11/20 19:00', '2003/1/10 20:00', '2005/7/19 13:00']   
end_time:  ['1994/2/28 23:00', '1997/4/29 11:00', '1997/6/29 00:00', '1997/7/29 15:00', '2001/12/27 09:00', '2002/11/15 10:00', '2003/1/3 10:00', '2005/7/16 01:00', '2022/5/4 23:00']   
invalid_times:  24344   ratio: 0.09779141794342366   
<img src='YL.png'> 

station_index:  
[[78, 182], [79, 168], [81, 162], [80, 199], [120, 154], [96, 202], [101, 173], [195, 154], [130, 181], [105, 169], [59, 170], [171, 171], [182, 270], [98, 221], [128, 146], [137, 78], [83, 60], [168, 100]]


HK location 22.1～22.7 113.8～114.6  
Spatial granularity: 0.01  800x600  
Temporal granularity: hour  2000~2021  

1. station spatial distribution  